### DISCUSSION TITLE
Supporting @vitejs/plugin-legacy

### DISCUSSION THREAD

**Discussion by johnRivs** [frandiox/vite-ssr#58] [Ideas]
I think this package could support the plugin legacy (or anything that provides multiple outputs for that matter) by making `build/index.ts` treat `clientResult` as an array.
![](https://i.imgur.com/KSPOiPm.png)

Seems to work fine but I'm sure you have more insight into how to properly do this.
![](https://i.imgur.com/NP28hCq.png)

### Why didn't I just submit a PR?
I'm making this post cause I can't figure out how to PR this 😂 I'm not familiar with TS, cloned the repo, ran `npm run dev` and `npm run build` but this is what I got
![](https://i.imgur.com/aX3C41X.png)
![](https://i.imgur.com/vD1rnPj.png)

**Comment by frandiox** [comment]
About the legacy build, I've tried it and the `clientResult` seems to be an object, not an array. It simply adds a new file in `clientResult.output` array.

I'm not sure how to use those new generated files, though 🤔 

About running the repo, you need to install deps using `yarn`, then `yarn dev` at the root. After that, run `yarn dev:ssr` in any of the examples.

Not sure if it works on windows, though, due to `rm -rf` command in some scripts.

**Reply by johnRivs** [reply]
> About the legacy build, I've tried it and the clientResult seems to be an object, not an array. It simply adds a new file in clientResult.output array.

Weird. When I use the legacy plugin:
```js
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import viteSSR from 'vite-ssr/plugin.js'
import legacy from '@vitejs/plugin-legacy'

export default defineConfig({
    plugins: [
        viteSSR(),
        legacy({
            targets: ['defaults']
        }),
        vue(),
    ]
})
```

and I log `clientResult`
```js
const clientResult = (await vite_1.build(clientBuildOptions));
console.log(clientResult);
const indexHtml = clientResult.output.find((file) => file.type === 'asset' && file.fileName === 'index.html');
```

what I get is this:
![](https://i.imgur.com/qrrmOnr.png)

If I don't use the plugin, I too get the object:
![](https://i.imgur.com/3UJUlov.png)

That should clarify what my original post means to do.

**Reply by frandiox** [reply]
🤯  This is bizarre. I always get it like this:

![image](https://user-images.githubusercontent.com/1634092/124393646-648f0880-dd36-11eb-8bbb-3aa39086f798.png)


I'm using Vite 2.3.2 here but I don't think they've changed the output in any recent version 🤔 

--
The code is:

```js
  const clientResult = (await build(clientBuildOptions)) as RollupOutput
  console.dir(clientResult, { depth: 1 })
```

**Reply by johnRivs** [reply]
I was on 2.3.8 but I switched to 2.3.2 and I still get the array. What I'm thinking is, it shouldn't matter cause the line of code I suggested translates to "if it's an array, leave it but if it isn't, wrap it in an array" and then the rest of the file works on the array.

In other words, whether the result is `{ output: [] }` or `[{ output: [] }, { output: [] }]` the file will be able to handle it thanks to that line.

**Reply by frandiox** [reply]
Yeah but I don't know what to do if it's an array at the top level. For each clientResult... do what? There can be only 1 index.html for the SSR build 🤔 
Or only one of those clientResults contain index.html?

**Reply by johnRivs** [reply]
I got a single index.html:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  <script type="module" crossorigin src="/assets/index.6f16668b.js"></script>
  <link rel="modulepreload" href="/assets/vendor.9a0c5e08.js">
</head>
  <body>
    <div id="app"></div>
    
  
<script nomodule>!function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",(function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()}),!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();</script>
  <script nomodule src="/assets/polyfills-legacy.017ef256.js"></script>
  <script nomodule id="vite-legacy-entry" data-src="/assets/index-legacy.e6845c1e.js">System.import(document.getElementById('vite-legacy-entry').getAttribute('data-src'))</script>
</body>
</html>
```

**Reply by johnRivs** [reply]
Managed to run the repo (mkdir needed quotes for the path) and got `[ { output: [Array] }, { output: [Array] } ]` and 1 single index.html like I mentioned yesterday. Are you sure you added the plugin? That made the difference for me. I ran `yarn run dev` at the root and `yarn run build` in the vue example https://github.com/johnRivs/vite-ssr/commit/40242fd138d7ff0ef9e8ffa5545f651fd4929844 Should log the same thing to the console.
![](https://i.imgur.com/QTb4bEU.png)

**Reply by frandiox** [reply]
Hey! Yeah I'm pretty sure I added the plugin, in fact you can see in the image I posted that it generated `assets/index.xxxx.legacy.js`, which is a result of the plugin.

In any case, I'll check your repo later when I have more time!

**Reply by johnRivs** [reply]
I'm blind 😂 Thanks for taking the time man.

**Reply by frandiox** [reply]
Related issue #63 

I'm not sure how the `ssr-manifest.json` works with this plugin. I can see 2 are generated so only the last one prevails (it includes both legacy and modern references).
